activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -662,14 +663,24 @@ def left_outer_joins!(*args) # :nodoc:
     #
     # If the condition is any blank-ish object, then #where is a no-op and returns
     # the current relation.
-    def where(opts = :chain, *rest)
-      if :chain == opts
+    def where(opts = :chain, *rest, &block)
+      new_scope = if :chain == opts && block_given?
+        self
+      elsif :chain == opts
         WhereChain.new(spawn)
       elsif opts.blank?
         self
       else
         spawn.where!(opts, *rest)
       end
+
+      if block_given?
```

Comment:
Having this `block_given?` check here seems to imply that there are other argument options that accept a block.

Such as:
```ruby
Foo.where(bar: 'blah') do |foo|
  foo.thing.eq('something')
end
```

I don't think this is something that you actually support, is it? I don't see any tests for that. It makes me feel like this should be on line 668 instead of the `self`.

Reply:
It's supported but just not to produce unexpected results. If you were to pass a hash and a block you would expect both to be taken into account. Though I think it's a bit ugly and I'd rather throw an exception.

Reply:
I think people should do one or the other, and that it would make more sense if lines 678-680 were cut and pasted into the `if` on 668 (removing the need for the `new_scope` variable.

```ruby
    def where(opts = :chain, *rest, &block)
      if :chain == opts && block_given?
        composer = QueryComposer.new(model)
        constraints = composer.instance_exec(composer, &block)
        spawn.where!(constraints)
      elsif :chain == opts
        WhereChain.new(spawn)
      elsif opts.blank?
        self
      else
        spawn.where!(opts, *rest)
      end
    end
```

Reply:
Introducing Arel as a public API should be treated as an _advanced feature_ and it would be helpful if the API communicated this.

Extending `where` to support multiple, different syntaxes feels a _little_ overloaded. The potential of using both in a single method call is potentially pretty confusing, as well.

Proposal: what if we moved this interface into `WhereChain` instead? Like so:
```ruby
Post.where.sql { |post| post[:created_at].lteq(Time.current) }
```